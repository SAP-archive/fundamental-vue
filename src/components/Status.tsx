import { Component, DefaultSlot, Prop, Base } from '@/core';

const typeMapping = {
    warning: 'Warning',
    error: 'Error',
    success: 'Success',
};

const statusIconMapping = {
    available: 'Available',
    away: 'Away',
    busy: 'Busy',
    offline: 'Offline',
};

type StatusType = keyof (typeof typeMapping);
const StatusTypes = Object.keys(typeMapping) as StatusType[];

type StatusIconType = keyof (typeof statusIconMapping);
const StatusIconTypes = Object.keys(statusIconMapping) as StatusIconType[];

interface Props {
    icon?: string | null;
    statusIcon?: StatusIconType;
    type?: StatusType | null;
}

@Component('Status')
@DefaultSlot('Text displayed inside the status.')
export class Status extends Base<Props> {
    @Prop('built-in status icon type', { acceptableValues: StatusIconTypes, type: String, default: null })
    public statusIcon!: StatusIconType;

    @Prop('status type', { acceptableValues: StatusTypes, type: String, default: null })
    public type!: StatusType | null;

    @Prop('icon displayed in the status indicator', { type: String, default: null })
    public icon!: string | null;

    public render() {
        return <span class={this.classes}>{this.$slots.default}</span>;
    }

    private get iconClasses() {
        const icon = this.icon;
        if (icon == null) {
            return {};
        }
        return {
            [`sap-icon--${icon}`]: true,
        };
    }

    private get classes() {
        return {
            'fd-status-label': true,
            'fd-status-label fd-status-label--success': this.type === 'success',
            'fd-status-label fd-status-label--warning': this.type === 'warning',
            'fd-status-label fd-status-label--error': this.type === 'error',

            'fd-status-label fd-status-label--available': this.statusIcon === 'available',
            'fd-status-label fd-status-label--away': this.statusIcon === 'away',
            'fd-status-label fd-status-label--busy': this.statusIcon === 'busy',
            'fd-status-label fd-status-label--offline': this.statusIcon === 'offline',
            ...this.iconClasses,
        };
    }
}
