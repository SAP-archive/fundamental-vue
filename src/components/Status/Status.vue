<template>
  <span :class="classes">
    <slot />
  </span>
</template>

<script>
const typeMapping = {
  warning: "warning",
  error: "error",
  success: "success"
};

const statusIconMapping = {
  available: "available",
  away: "away",
  busy: "busy",
  offline: "offline"
};

const StatusTypes = Object.keys(typeMapping);
const isStatusType = value => StatusTypes.indexOf(value) >= 0;

const StatusIconTypes = Object.keys(statusIconMapping);
const isStatusIcon = value => StatusIconTypes.indexOf(value) >= 0;

export default {
  name: "FdStatus",
  props: {
    statusIcon: { type: String, default: null, validator: isStatusIcon },
    type: { type: String, default: null, validator: isStatusType },
    icon: { type: String, default: null }
  },
  computed: {
    iconClasses() {
      const icon = this.icon;
      if (icon == null) {
        return {};
      }
      return {
        [`sap-icon--${icon}`]: true
      };
    },
    classes() {
      return {
        "fd-status-label": true,
        "fd-status-label--success": this.type === "success",
        "fd-status-label--warning": this.type === "warning",
        "fd-status-label--error": this.type === "error",
        "fd-status-label--available": this.statusIcon === "available",
        "fd-status-label--away": this.statusIcon === "away",
        "fd-status-label--busy": this.statusIcon === "busy",
        "fd-status-label--offline": this.statusIcon === "offline",
        ...this.iconClasses
      };
    }
  }
};
</script>
