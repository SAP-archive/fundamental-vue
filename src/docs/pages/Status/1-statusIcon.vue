<title>Status Indicator Label with build in status icons</title>
<template>
  <div class="status">
    <FdStatus statusIcon="available">Available</FdStatus>
    <FdStatus statusIcon="away">Away</FdStatus>
    <FdStatus statusIcon="busy">Busy</FdStatus>
    <FdStatus statusIcon="offline">Appear Offline</FdStatus>
  </div>
</template>
