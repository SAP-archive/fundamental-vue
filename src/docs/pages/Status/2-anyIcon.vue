<title>Status Indicator Label with any icons</title>

<template>
  <div class="status">
    <FdStatus icon="history">Custom Icon</FdStatus>
    <FdStatus icon="message-success">Success</FdStatus>
    <FdStatus icon="message-warning">Warning</FdStatus>
    <FdStatus icon="message-error">Error</FdStatus>
  </div>
</template>
