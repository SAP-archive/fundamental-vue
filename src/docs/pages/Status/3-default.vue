<title>Status Indicator Label with symentic colors</title>

<template>
  <div class="status">
    <FdStatus>Default</FdStatus>
    <FdStatus type="success">Success</FdStatus>
    <FdStatus type="warning">Warning</FdStatus>
    <FdStatus type="error">Error</FdStatus>
  </div>
</template>

<style scoped>
.status {
  text-align: center;
}
.status > span {
  margin-right: 50px;
}
</style>
