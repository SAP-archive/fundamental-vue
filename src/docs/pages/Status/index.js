export default {
  relatedComponents: ["fd-status"]
};
