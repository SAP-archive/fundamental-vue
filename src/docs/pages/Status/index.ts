import { ExampleCollectionFunction } from '../types';

export const plugin: ExampleCollectionFunction = ({ Status }) => {
    return { componentStatus: 'stable', icon: 'order-status', relatedComponents: [Status] };
};
