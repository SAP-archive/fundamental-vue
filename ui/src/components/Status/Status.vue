<template>
  <span :class="classes">
    <slot />
  </span>
</template>

<script lang="ts">
import Vue from "vue";

const typeMapping = {
  warning: "warning",
  error: "error",
  success: "success"
};

const statusIconMapping = {
  available: "available",
  away: "away",
  busy: "busy",
  offline: "offline"
};

const StatusTypes = Object.keys(typeMapping);
const isStatusType = (value: string) => StatusTypes.indexOf(value) >= 0;

const StatusIconTypes = Object.keys(statusIconMapping);
const isStatusIcon = (value: string) => StatusIconTypes.indexOf(value) >= 0;

export default Vue.extend({
  name: "FdStatus",
  props: {
    statusIcon: { type: String, default: null, validator: isStatusIcon },
    type: { type: String, default: null, validator: isStatusType },
    icon: { type: String, default: null }
  },
  computed: {
    iconClasses(): object {
      const icon = this.icon;
      if (icon == null) {
        return {};
      }
      return {
        [`sap-icon--${icon}`]: true
      };
    },
    classes(): object {
      return {
        "fd-status-label": true,
        "fd-status-label--success": this.type === "success",
        "fd-status-label--warning": this.type === "warning",
        "fd-status-label--error": this.type === "error",
        "fd-status-label--available": this.statusIcon === "available",
        "fd-status-label--away": this.statusIcon === "away",
        "fd-status-label--busy": this.statusIcon === "busy",
        "fd-status-label--offline": this.statusIcon === "offline",
        ...this.iconClasses
      };
    }
  }
});
</script>
