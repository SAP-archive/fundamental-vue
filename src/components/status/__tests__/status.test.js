import { mount } from '@vue/test-utils'
import Status from '../status.vue'

describe('Status test scripts', () => {
  test('Renders correctly', () => {
    const statusIcon = 'available'
    const type = 'success'
    const status = mount(Status, {
      propsData: {
        statusIcon,
        type
      }
    })
    expect(status.element).toMatchSnapshot()
  })

  test('Default slot rendering test', () => {
    const slotText = 'Status Slot Text'
    const status = mount(Status, {
      slots: {
        default: slotText
      }
    })
    expect(status.text()).toBe(slotText)
  })

  test('Status indicator with built in status icon test', () => {
    const statusIcon = 'offline'
    const status = mount(Status, {
      propsData: {
        statusIcon
      }
    })
    expect(status.classes(`fd-status-label--${statusIcon}`)).toBe(true)
  })

  test('Status indicator with semantic colors', () => {
    const type = 'warning'
    const status = mount(Status, {
      propsData: {
        type
      }
    })
    expect(status.classes(`fd-status-label--${type}`)).toBe(true)
  })
})
