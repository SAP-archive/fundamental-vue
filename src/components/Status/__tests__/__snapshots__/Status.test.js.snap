// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status test scripts Renders correctly 1`] = `
<span
  class="fd-status-label fd-status-label--success fd-status-label--available"
/>
`;
